drivers/led/is31fl3194.c

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2024 Arduino SA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT issi_is31fl3194
+
+/**
+ * @file
+ * @brief IS31FL3194 LED driver
+ *
+ * The IS31FL3194 is a 3-channel LED driver that communicates over I2C.
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <zephyr/dt-bindings/led/led.h>
+
+LOG_MODULE_REGISTER(is31fl3194, CONFIG_LED_LOG_LEVEL);
+
+#define IS31FL3194_PROD_ID_REG	0x00
+#define IS31FL3194_CONF_REG	0x01
+#define IS31FL3194_OUT_EN_REG	0x02
+#define IS31FL3194_CURRENT_REG	0x03
+#define IS31FL3194_OUT1_REG	0x10
+#define IS31FL3194_OUT2_REG	0x21
+#define IS31FL3194_OUT3_REG	0x32
+#define IS31FL3194_UPDATE_REG	0x40
+
+#define IS31FL3194_PROD_ID_VAL		0xce
+#define IS31FL3194_CONF_ENABLE		0x01
+#define IS31FL3194_CONF_RGB_EN		0x04
+#define IS31FL3194_OUT_EN_OUT1		0x01
+#define IS31FL3194_OUT_EN_OUT2		0x02
+#define IS31FL3194_OUT_EN_OUT3		0x04
+#define IS31FL3194_UPDATE_VAL		0xc5
+
+#define IS31FL3194_CHANNEL_COUNT 3
+
+static const uint8_t led_channels[] = {
+	IS31FL3194_OUT1_REG,
+	IS31FL3194_OUT2_REG,
+	IS31FL3194_OUT3_REG
+};
+
+struct is31fl3194_config {
+	struct i2c_dt_spec bus;
+	uint8_t num_leds;
+	const struct led_info *leds_info;
+	const uint8_t *current_limits;
+};
+
+static const struct led_info *is31fl3194_led_to_info(const struct is31fl3194_config *config,
+						     uint32_t led)
+{
+	if (led < config->num_leds) {
+		return &config->leds_info[led];
+	}
+
+	return NULL;
+}
+
+static int is31fl3194_get_info(const struct device *dev,
+			       uint32_t led,
+			       const struct led_info **info)
+{
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *led_info = is31fl3194_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t num_colors,
+			     const uint8_t *color)
+{
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *led_info = is31fl3194_led_to_info(config, led);
+	int ret;
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (led_info->num_colors != 3) {
+		return -ENOTSUP;
+	}
+	if (num_colors != 3) {
```

Comment:
Do I? This API does not provide the "start offset" so you are not free to edit a single specific channel; I could only allow editing subsets from index 0 onwards. I feel this would be an "unintended feature" of the API :slightly_smiling_face: 

This also raises a related question: what is the data ordering here? How should color mapping be handled? AFAICS:
* "led" drivers _ignore_ color mapping info and leave it to the user to provide input in the correct order; 
* "led_strip" drivers map RGB-ordered data (from `struct led_rgb`) to the appropriate channels using the color mapping info.

I did implement the latter without much thinking about it, since I assumed that was the intention of the `color-mapping` info in the DTS. Why should the user have to deal with this in their app?

Reply:
> Do I? This API does not provide the "start offset" so you are not free to edit a single specific channel; I could only allow editing subsets from index 0 onwards. I feel this would be an "unintended feature" of the API 🙂

Yes you are right. It is actually the purpose of the `->write_channels()` API (not implemented in your driver) to allow specific channels to be updated.

> 
> This also raises a related question: what is the data ordering here? How should color mapping be handled? AFAICS:
> 
>     * "led" drivers _ignore_ color mapping info and leave it to the user to provide input in the correct order;

Yes it is assumed that all the colors of a single LED are passed to `->set_color()` following the order defined by the `color-mapping` DT property.

> 
>     * "led_strip" drivers map RGB-ordered data (from `struct led_rgb`) to the appropriate channels using the color mapping info.

Yes. And it is something I'd like to see unified with the idea of merging the led and led_strip subsystems.

> 
> I did implement the latter without much thinking about it, since I assumed that was the intention of the `color-mapping` info in the DTS. Why should the user have to deal with this in their app?

And that's correct :)